src/Swarm/TUI/View.hs

```diff
@@ -844,21 +848,25 @@ drawWorld g =
       let w = ctx ^. availWidthL
           h = ctx ^. availHeightL
           ixs = range (viewingRegion g (fromIntegral w, fromIntegral h))
-      render . vBox . map hBox . chunksOf w . map (drawLoc g) $ ixs
+      render . vBox . map hBox . chunksOf w . map (drawLoc showRobots g) $ ixs
 
 -- | Render the 'Display' for a specific location.
-drawLoc :: GameState -> W.Coords -> Widget Name
-drawLoc g = renderDisplay . displayLoc g
+drawLoc :: Bool -> GameState -> W.Coords -> Widget Name
+drawLoc showRobots g = renderDisplay . displayLoc showRobots g
 
 -- | Get the 'Display' for a specific location, by combining the
 --   'Display's for the terrain, entity, and robots at the location.
-displayLoc :: GameState -> W.Coords -> Display
-displayLoc g coords =
-  sconcat . NE.fromList $
-    [terrainMap M.! toEnum (W.lookupTerrain coords (g ^. world))]
-      ++ maybeToList (displayForEntity <$> W.lookupEntity coords (g ^. world))
-      ++ map (view robotDisplay) (robotsAtLocation (W.coordsToLoc coords) g)
+displayLoc :: Bool -> GameState -> W.Coords -> Display
+displayLoc showRobots g coords =
+  sconcat $ terrain NE.:| entity <> robots
  where
+  terrain = terrainMap M.! toEnum (W.lookupTerrain coords (g ^. world))
+  entity = maybeToList (displayForEntity <$> W.lookupEntity coords (g ^. world))
+  robots =
+    if showRobots
+      then map (view robotDisplay) (robotsAtLocation (W.coordsToLoc coords) g)
+      else []
```

Comment:
Looks much cleaner than my original code, thanks. :smile:

Reply:
STAN shows me warnings about unsafe functions in HLS and in this case, it got me thinking about refactoring to use the `NonEmpty` list constructor instead of `fromList`.

I usually just ignore them (they are not intrusive) but here it worked nicely. 😁